### ISSUE TITLE
Allow capturing minidumps programatically

### ISSUE THREAD

**Issue by bruno-garcia** [getsentry/sentry-native#1050]
A customer who has their own crash handler mechanism would like to upload minidumps they capture by themselves.
One of the solutions proposed was using HTTP and [uploading them to our `minidump` endpoint](https://docs.sentry.io/platforms/native/guides/minidumps/).

Since that means having to deal with retries, managing files on disk etc. I suggesting using one of our SDKs to manage that.

Our envelope format allows attachment, including type `minidump`. Which should trigger symbolication.

While we might not have a `capture_minidump` yet, the suggestion is to create this function so the SDK takes over the job of establishing the connection to Sentry. And deleting the file upon completion.

**Comment by kahest** [comment]
@bruno-garcia you mention

> Since that means having to deal with retries, managing files on disk etc.

`sentry-native` currently doesn't do this, and there's no synchronous behaviour between client and transport. so if that's the expectation it would require additional effort than what is suggested here

**Comment by PlasmaDev5** [comment]
Wrote up some initial notes on this
[HandleCustomMinidumps.pdf](https://github.com/user-attachments/files/17335867/HandleCustomMinidumps.pdf)

**Comment by vaind** [comment]
> [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1633368?s=64&amp;u=856a9e52d4cdc92eac2ff9e1f97900df39ae5e4d&amp;v=4">@bruno-garcia](https://github.com/bruno-garcia) you mention
> 
> > Since that means having to deal with retries, managing files on disk etc.
> 
> `sentry-native` currently doesn't do this, and there's no synchronous behaviour between client and transport. so if that's the expectation it would require additional effort than what is suggested here

I don't think we need that - IMO, it would be fine to just process all minidumps at once, which moves them over to the envelope cache dir and transport handles retries. Also, we can probably set any large timeout to flush() so it runs until the server accepts everything.

**Comment by vaind** [comment]
> A customer who has their own crash handler mechanism would like to upload minidumps they capture by themselves.

do they want to deal with sentry-native API, building an executable to run this? IDK but build a simple script that runs HTTP requests to our backend sounds a lot simpler.

**Comment by PlasmaDev5** [comment]
> do they want to deal with sentry-native API, building an executable to run this? IDK but build a simple script that runs HTTP requests to our backend sounds a lot simpler.

From my understanding they are already making use of sentry-native for the use in question this would be a more direct way of achieving there goal. I imagine @bruno-garcia can give more insight on this

**Comment by bruno-garcia** [comment]
> > do they want to deal with sentry-native API, building an executable to run this? IDK but build a simple script that runs HTTP requests to our backend sounds a lot simpler.
> 
> From my understanding they are already making use of sentry-native for the use in question this would be a more direct way of achieving there goal. I imagine [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1633368?s=64&amp;u=856a9e52d4cdc92eac2ff9e1f97900df39ae5e4d&amp;v=4">@bruno-garcia](https://github.com/bruno-garcia) can give more insight on this

That's right, they already have `sentry-native` on the app. And my understanding is that in the future they might switch to our own backends and drop their custom minidump collecting logic.

**Comment by supervacuus** [comment]
> > [<img alt="" width="16" height="16" src="https://avatars.githubusercontent.com/u/1633368?s=64&amp;u=856a9e52d4cdc92eac2ff9e1f97900df39ae5e4d&amp;v=4">@bruno-garcia](https://github.com/bruno-garcia) you mention
> > > Since that means having to deal with retries, managing files on disk etc.
> > 
> > 
> > `sentry-native` currently doesn't do this, and there's no synchronous behaviour between client and transport. so if that's the expectation it would require additional effort than what is suggested here
> 
> I don't think we need that - IMO, it would be fine to just process all minidumps at once, which moves them over to the envelope cache dir and transport handles retries. Also, we can probably set any large timeout to flush() so it runs until the server accepts everything.

To be clear, because you mention "transport handles retries" again in your response, this is currently _not_ happening. The transport only handles rate-limiting returns (and delays according to responses from the server) but not intermittent network downtime (or any other network failure). There is no exponential backoff or similar strategy in place.

So, while removing source minidumps concerning transferring their data into an envelope attachment is safe, once the sender thread picks up the envelope, it is gone. 

Flushing is only related to the topic as it persists the inflight items of the send queue. Still, the sender thread is not transactional in any way (i.e., if it fails to send beyond `429`, `Retry-After`, `x-sentry-rate-limits`, it will dequeue all remaining items in the queue, which might equally fail to send).

**Comment by bruno-garcia** [comment]
Understood.

Retries I meant is more about offline mode. Since we don't have retries built-in as in "Failed to send to Sentry, lets retry" as this can result in weird DDoS scenarios, and clients stuck for some reason retrying forever if they have some connectivity issue.
So seems like we need to at least keep the minidump on disk if the device is offline. But since this is mainly used on desktop/server, this is very much a nice to have feature.

In conclusion, we can forget about 'retries'. Rate limiting is the value one gets from having to try to write this by themselves.
